src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -61,38 +123,47 @@ object EvalContext {
   }
 }
 
-trait NumericConversion[T] extends Serializable {
+sealed trait NumericConversion[T, U <: ToBoxed[T]] extends Serializable {
   def to(numeric: Any): T
+  def to(c: Code[java.lang.Number]): CM[Code[U#T]]
```

Comment:
`ToBoxed[T]` is a lightweight way to ensure we define the boxed version of `T` in one place and are consistent about it. Does this already exist in Scala somewhere?